File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/fixtures.ts
/**
 * Campaign-scoped deterministic fixture seeding: only the campaign DM may
 * seed the canonical fixture; any campaign member (including the DM) may
 * read the seeded state.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { authenticate, isActor, isApiResult, findCampaign, requireParticipant } from './shared.ts';

const CANONICAL_FIXTURE_ID = 'canonical-v1';

const CANONICAL_FIXTURE_STATE = {
  fixture_id: CANONICAL_FIXTURE_ID,
  status: 'seeded',
  characters: [
    { character_id: 'fixture-hero', name: 'Ari', class: 'fighter' },
    { character_id: 'fixture-mage', name: 'Bea', class: 'wizard' },
  ],
  story: 'The lantern is lit.',
  event_ids: ['fixture-event-1', 'fixture-event-2'],
} as const;

export function seedFixture(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may seed fixture state' } };
  }

  const fixtureId = body?.fixture_id;
  if (typeof fixtureId !== 'string' || fixtureId.length === 0 || fixtureId !== CANONICAL_FIXTURE_ID) {
    return { status: 400, body: { error: "fixture_id must be exactly 'canonical-v1'" } };
  }

  const existing = db
    .prepare('SELECT fixture_id FROM play_campaign_fixture_seeds WHERE campaign_id = ? AND fixture_id = ?')
    .get(campaignId, fixtureId);

  if (!existing) {
    db.prepare('INSERT INTO play_campaign_fixture_seeds (campaign_id, fixture_id) VALUES (?, ?)').run(
      campaignId,
      fixtureId,
    );
    return { status: 201, body: CANONICAL_FIXTURE_STATE as unknown as JsonValue };
  }

  return { status: 200, body: CANONICAL_FIXTURE_STATE as unknown as JsonValue };
}

export function getFixtureState(authHeader: string | undefined, campaignId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const existing = db
    .prepare('SELECT fixture_id FROM play_campaign_fixture_seeds WHERE campaign_id = ? AND fixture_id = ?')
    .get(campaignId, CANONICAL_FIXTURE_ID);
  if (!existing) {
    return { status: 404, body: { error: 'fixture not seeded for this campaign' } };
  }

  return { status: 200, body: CANONICAL_FIXTURE_STATE as unknown as JsonValue };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T002352Z_pi_kimi-k2p7-code_python-django/dndsite/views/fixtures.py
"""Campaign-scoped deterministic fixture seeding views."""

from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from ..db import _get_fixture_seed, _seed_fixture, db_conn
from ..http import bad_request, not_found, parse_json, require_method
from .common import require_play_campaign_dm_owner, require_play_campaign_member_or_owner


CANONICAL_FIXTURE_ID = "canonical-v1"


def _build_fixture_response(fixture_id):
    """Return the deterministic fixture state for the seeded fixture."""
    return {
        "fixture_id": fixture_id,
        "status": "seeded",
        "characters": [
            {"character_id": "fixture-hero", "name": "Ari", "class": "fighter"},
            {"character_id": "fixture-mage", "name": "Bea", "class": "wizard"},
        ],
        "story": "The lantern is lit.",
        "event_ids": ["fixture-event-1", "fixture-event-2"],
    }


@csrf_exempt
def fixture_seeds(request, id):
    """POST /v1/play/campaigns/{id}/fixture-seeds"""
    bad = require_method(request, "POST")
    if bad:
        return bad

    actor, campaign, err = require_play_campaign_dm_owner(request, id)
    if err is not None:
        return err

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")

    try:
        fixture_id = body["fixture_id"]
        if not isinstance(fixture_id, str) or fixture_id != CANONICAL_FIXTURE_ID:
            raise ValueError
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    with db_conn() as conn:
        existing = _get_fixture_seed(conn, id)
        if existing is not None:
            return JsonResponse(_build_fixture_response(existing["fixture_id"]), status=200)
        _seed_fixture(conn, id, fixture_id)

    return JsonResponse(_build_fixture_response(fixture_id), status=201)


@csrf_exempt
def fixture_state(request, id):
    """GET /v1/play/campaigns/{id}/fixture-state"""
    bad = require_method(request, "GET")
    if bad:
        return bad

    actor, campaign, err = require_play_campaign_member_or_owner(request, id)
    if err is not None:
        return err

    with db_conn() as conn:
        existing = _get_fixture_seed(conn, id)
        if existing is None:
            return not_found("fixture state not found")

    return JsonResponse(_build_fixture_response(existing["fixture_id"]))
